lib/src/model/tournament/tournament_controller.dart

```diff
@@ -0,0 +1,150 @@
+import 'package:lichess_mobile/src/model/tournament/tournament.dart';
+import 'package:lichess_mobile/src/model/tournament/tournament_repository.dart';
+import 'package:lichess_mobile/src/network/socket.dart';
+import 'package:logging/logging.dart';
+import 'package:riverpod_annotation/riverpod_annotation.dart';
+
+part 'tournament_controller.freezed.dart';
+part 'tournament_controller.g.dart';
+
+final _logger = Logger('TournamentController');
+
+@riverpod
+class TournamentController extends _$TournamentController {
+  StreamSubscription<SocketEvent>? _socketSubscription;
+
+  late final SocketClient _socketClient;
+
+  // If we join/leave to often, the server blocks us from joining, see [TournamentMe.pauseDelay].
+  // However, there's no "reload" event from the socket once we're able to join again,
+  // so we manually have to set this timer to schedule a reload once we can join again.
+  Timer? _pauseDelayTimer;
+
+  @override
+  Future<TournamentState> build(TournamentId id) async {
+    final state = await _loadTournament(id, standingsPage: 1);
+
+    final socketPool = ref.watch(socketPoolProvider);
+    _socketClient = socketPool.open(Uri(path: '/tournament/$id/socket/v6'));
+    ref.onDispose(() {
+      _socketSubscription?.cancel();
+      _pauseDelayTimer?.cancel();
+    });
+
+    _socketSubscription?.cancel();
+    _socketSubscription = _socketClient.stream.listen(_handleSocketEvent);
+
+    return state;
+  }
+
+  Future<TournamentState> _loadTournament(TournamentId id, {required int standingsPage}) async {
+    final tournament = await ref
+        .read(tournamentRepositoryProvider)
+        .getTournament(id, standingsPage: standingsPage);
+    return TournamentState(tournament: tournament, standingsPage: standingsPage);
+  }
+
+  void loadNextStandingsPage() {
+    if (state.hasValue) {
+      _reload(standingsPage: state.requireValue.standingsPage + 1);
+    }
+  }
+
+  void loadPreviousStandingsPage() {
+    if (state.hasValue) {
+      _reload(standingsPage: state.requireValue.standingsPage - 1);
+    }
+  }
+
+  void loadFirstStandingsPage() {
+    _reload(standingsPage: 1);
+  }
+
+  int _pageOf(int page) => page ~/ kStandingsPageSize + 1;
+
+  void loadLastStandingsPage() {
+    if (state.hasValue) {
+      _reload(standingsPage: _pageOf(state.requireValue.tournament.nbPlayers));
+    }
+  }
+
+  void jumpToMyPage() {
+    if (state.valueOrNull?.tournament.me != null) {
+      _reload(standingsPage: _pageOf(state.requireValue.tournament.me!.rank));
+    }
+  }
+
+  Future<void> _reload({required int standingsPage}) async {
+    _logger.fine('Refreshing tournament standings page $standingsPage');
+
+    if (!state.hasValue) {
+      return;
+    }
+
+    final tournament = await ref
+        .read(tournamentRepositoryProvider)
+        .reload(state.requireValue.tournament, standingsPage: standingsPage);
+
+    if (tournament.me?.pauseDelay != null) {
+      _pauseDelayTimer = Timer(tournament.me!.pauseDelay!, () {
```

Comment:
Always cancel the timer before building a new one.

Reply:
ah, coming from C++ it's weird that there are no destructors that take care of this stuff, will try to remember this next time :D